modules/finality-verifier/src/lib.rs

```diff
@@ -64,17 +65,24 @@ pub mod pallet {
 		/// The pallet which we will use as our underlying storage mechanism.
 		type HeaderChain: HeaderChain<<Self::BridgedChain as Chain>::Header, DispatchError>;
 
+		/// The type of ancestry proof used by the pallet.
+		///
+		/// Will be used by the ancestry checker to verify that the header being finalized is
+		/// related to the best finalized header in storage.
+		type AncestryProof: Parameter + Size;
+
 		/// The type through which we will verify that a given header is related to the last
 		/// finalized header in our storage pallet.
-		type AncestryChecker: AncestryChecker<
-			<Self::BridgedChain as Chain>::Header,
-			Vec<<Self::BridgedChain as Chain>::Header>,
-		>;
+		type AncestryChecker: AncestryChecker<<Self::BridgedChain as Chain>::Header, Self::AncestryProof>;
 
-		/// The maximum length of headers we can have in a single ancestry proof. This prevents
-		/// unbounded iteration when verifying proofs.
+		/// The maximum length of elements we can have in a single ancestry proof.
+		///
+		/// It is an optional field since some ancestry proof types may require iterating through
+		/// the proof during verification. If left unbounded, this could be problematic. On the
+		/// other hand, some proof structures may not care about a maximum length, in which case
+		/// this does not need to be used.
 		#[pallet::constant]
-		type MaxHeadersInSingleProof: Get<<Self::BridgedChain as Chain>::BlockNumber>;
+		type MaxElementsInSingleProof: Get<Option<u32>>;
```

Comment:
IMHO if we were to create a proper abstraction this should be function from the `AncestryChecker` trait and it should simply receive the `ancestry_proof` parameter and it will know the type, so can easily verify whatever is needed (without `Size::size_hint` hacks).

Reply:
I don't think that limiting the number of elements in the proof should be part of the trait's responsibility. It should be done as part of wiring up the runtime.

I have changed the `AncestryChecker` trait to provide a method for easily getting the proof size, let me know if that works for you.

Reply:
It's better than `size_hint` now, but I still don't think it's the right way. Since the `proof` is already abstract the notion of `proof_size` is completely ambiguous.

What `proof_size` does `frame_system` storage proof or MMR proof have? Is it the number of proof nodes or maybe total encoded bytes length, or maybe even it's always 1 cause it contains only one header, given that `Vec<Header>` `proof_size` is defined as the length of the vector.

Delegation to the abstract `AncestryChecker` will be required for weight computation anyway, so IMHO either we do a proper abstraction, where we don't make assumptions about what properties the "ancestry proof" has (i.e. `proof_size`) or we avoid introducing this abstraction at all, and let's simply take `Vec<Header>` for now and verify by calling into a regular function not a trait (my preference actually).

I feel this is all moot anyway, since the code we talk about here should be removed ASAP and replaced with storage proof verification (as described in #628)

Reply:
> It's better than size_hint now, but I still don't think it's the right way. Since the proof is already abstract the notion of proof_size is completely ambiguous.

While I figured it would be up to the implementer of whatever type of ancestry checker to decide how they want to size their proof, I agree with you it's not a great property to add directly to the ancestry checker.

> Delegation to the abstract AncestryChecker will be required for weight computation anyway, so IMHO either we do a proper abstraction, where we don't make assumptions about what properties the "ancestry proof" has (i.e. proof_size) or we avoid introducing this abstraction at all, and let's simply take Vec<Header> for now and verify by calling into a regular function not a trait (my preference actually).

I like the flexibility given by the trait, so I'm going to stick with (roughly) the original - maybe removing the size checks since their meaning is ambiguous (🤠).

> I feel this is all moot anyway, since the code we talk about here should be removed ASAP and replaced with storage proof verification (as described in #628)

Do you mind expanding on this point here? I envisioned that the same pallet code could be used. The only difference would be that in the runtime we would need an implementation of `AncestryChecker` that used storage proofs.

Reply:
> maybe removing the size checks since their meaning is ambiguous

That's a step backwards and will cause problems when calculating weights, but okay, let's deal with it when we implement weights.

> Do you mind expanding on this point here? I envisioned that the same pallet code could be used. 

Yes, the same pallet will be used.

> The only difference would be that in the runtime we would need an implementation of AncestryChecker that used storage proofs.

No, that's not the only difference. The differences are:
1. Different `AncestryChecker` implementation
2. Different proof type (hence I suggested an associated type on `AncestryChecker` trait)
3. Different limiting mechanism, specific to `AncestryChecker` (hence suggested to simply move the check to the trait)
4. Different weight mechanism, specific to `AncestryChecker` (that would come next, but it would simply be another trait method on the trait).

TBH Given that we only plan to have one `AncestryChecker` anyway, introducing the trait is pre-mature generalisation (is that a thing even :D), hence the entire discussion.

That said, let's just merge it and start making some progress instead of discussing.

Reply:
> TBH Given that we only plan to have one AncestryChecker anyway, introducing the trait is pre-mature generalisation (is that a thing even :D), hence the entire discussion.

When I wrote this up I thought we'd have at least three `AncestryChecker` implementations, which is why it's traitified :)

> That said, let's just merge it and start making some progress instead of discussing.

👌